### PR TITLE
Improve SIFT for arm64/Apple silicon

### PR DESCRIPTION
**Pull request by Developer-Ecosystem-Engineering**
Reduce branch density by collapsing compares.

Performance improvements from 1.03 to 1.53 with existing tests

<cut/>

```
% ./modules/ts/misc/summary.py -u us -m median ./build-{3.4,improv-sift}/*.xml | grep SIFT
(x-factor)
detect::OCL_feature2d::(SIFT_DEFAULT, "cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png") 50256.333 48789.417 1.03
detect::OCL_feature2d::(SIFT_DEFAULT, "stitching/a3.png") 64591.083 52250.542 1.24
detect::OCL_feature2d::(SIFT_DEFAULT, "stitching/s2.jpg") 132094.667 99417.667 1.33
detect::feature2d::(SIFT_DEFAULT, "cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png") 77521.167 60114.083 1.29
detect::feature2d::(SIFT_DEFAULT, "stitching/a3.png") 70179.083 55914.917 1.26
detect::feature2d::(SIFT_DEFAULT, "stitching/s2.jpg") 137359.208 101112.250 1.36
detectAndExtract::OCL_feature2d::(SIFT_DEFAULT, "cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png") 99694.083 67476.833 1.48
detectAndExtract::OCL_feature2d::(SIFT_DEFAULT, "stitching/a3.png") 101392.958 71199.708 1.42
detectAndExtract::OCL_feature2d::(SIFT_DEFAULT, "stitching/s2.jpg") 203133.500 133395.083 1.52
detectAndExtract::feature2d::(SIFT_DEFAULT, "cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png") 98465.083 66492.083 1.48
detectAndExtract::feature2d::(SIFT_DEFAULT, "stitching/a3.png") 100793.042 68811.333 1.46
detectAndExtract::feature2d::(SIFT_DEFAULT, "stitching/s2.jpg") 203328.083 132630.917 1.53
extract::OCL_feature2d::(SIFT_DEFAULT, "cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png") 26723.042 19980.250 1.34
extract::OCL_feature2d::(SIFT_DEFAULT, "stitching/a3.png") 19655.625 15591.500 1.26
extract::OCL_feature2d::(SIFT_DEFAULT, "stitching/s2.jpg") 59479.917 44881.208 1.33
extract::feature2d::(SIFT_DEFAULT, "cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png") 27984.542 19793.625 1.41
extract::feature2d::(SIFT_DEFAULT, "stitching/a3.png") 21164.250 16235.250 1.30
extract::feature2d::(SIFT_DEFAULT, "stitching/s2.jpg")
extract::feature2d::(SIFT_DEFAULT, "stitching/s2.jpg") 58240.000 44191.833 1.32
```

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=ARMv7,ARMv8,Custom
buildworker:Custom=linux-3
build_image:Custom=ubuntu:18.04
CPU_BASELINE:Custom=AVX512_SKX
disable_ipp=ON
```

### CODE REVIEW

**Comment by Developer-Ecosystem-Engineering** [comment]
Investigating build failures

**Comment by tomoaki0705** [comment]
TIPS 1 : rather filtering with `grep`, I recommend to filter using `--gtest_filter` option that you pass to the gtest executable file

```
./opnecv_test_features2d --gtest_filter=*SIFT*
```

TIPS 2: use `-o markdown` option of `summary.py`

```
summary.py -o markdown -m median build-before.xml build-after.xml
```

Will give you something like this

```
Median
|Name of Test|build-before|build-after|build-before vs build-after(x-factor)|
|---|:-:|:-:|:-:|
|detect::OCL_feature2d::(SIFT_DEFAULT, "cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png")|50.256 ms|48.789417 ms|1.03|
|detect::OCL_feature2d::(SIFT_DEFAULT, "stitching/a3.png")|64.591083ms| 52.250542ms | 1.24 |
```

Which is WAY MORE READABLE on Github

Median
|Name of Test|build-before|build-after|build-before vs build-after(x-factor)|
|---|:-:|:-:|:-:|
|detect::OCL_feature2d::(SIFT_DEFAULT, "cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png")|50.256 ms|48.789 ms|1.03|
|detect::OCL_feature2d::(SIFT_DEFAULT, "stitching/a3.png")|64.591ms| 52.250ms | 1.24 |



Also, the commit message 
> only enable for NEON
is a red flag for me.

**Comment by Developer-Ecosystem-Engineering** [comment]
Thanks for the tips, will reformat.

> > only enable for NEON
> > is a red flag for me.

We tested full vector on coffeelake, marginal up (and down).
Tested 128 bit vector and saw even lower benefits. 

Since the benefit was clear on NEON, thats where we landed.

File: modules/features2d/src/sift.simd.hpp.orig [hunk 1]
```diff
@@ -0,0 +1,865 @@
+// This file is part of OpenCV project.
```
Comment (asmorkalov) [inline modules/features2d/src/sift.simd.hpp.orig:1]: I suppose that the file was added to repo by mistake.

File: modules/features2d/src/sift.simd.hpp [hunk 2]
```diff
@@ -450,31 +450,182 @@ class findScaleSpaceExtremaT
             const sift_wt* currptr = img.ptr<sift_wt>(r);
             const sift_wt* prevptr = prev.ptr<sift_wt>(r);
             const sift_wt* nextptr = next.ptr<sift_wt>(r);
+            int c = SIFT_IMG_BORDER;
 
-            for( int c = SIFT_IMG_BORDER; c < cols-SIFT_IMG_BORDER; c++)
+#if CV_NEON && !(DoG_TYPE_SHORT)
```
Comment (asmorkalov) [inline modules/features2d/src/sift.simd.hpp:455]: I propose to use CV_SIMD128 macro to guard the changes. The patch may be relevant for other architectures, besides Intel and ARM.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

**Comment by vpisarev** [comment]
@Developer-Ecosystem-Engineering, thank you for the contribution! This work on improving OpenCV@M1 performance is brilliant!

Note, however, that except for the kernels in DNN module, which are few and really critical, we do not accept native optimizations any longer. It would be just impossible for our tiny team to maintain all those branches. Please, rewrite the native NEON code using our universal intrinsics.

File: modules/features2d/src/sift.simd.hpp.orig (hunk 1 above)
Comment (Developer-Ecosystem-Engineering) [reply modules/features2d/src/sift.simd.hpp.orig:1]: Removed in the latest PR, apologies.

**Comment by Developer-Ecosystem-Engineering** [comment]
> @Developer-Ecosystem-Engineering, thank you for the contribution! This work on improving OpenCV@M1 performance is brilliant!
> 
> Note, however, that except for the kernels in DNN module, which are few and really critical, we do not accept native optimizations any longer. It would be just impossible for our tiny team to maintain all those branches. Please, rewrite the native NEON code using our universal intrinsics.

An updated patch is available with it rewritten.

File: modules/features2d/src/sift.simd.hpp (hunk 2 above)
Comment (Developer-Ecosystem-Engineering) [reply modules/features2d/src/sift.simd.hpp]: We previously had it using SIMD but the gains were not really showing as worth it so we constrained it.

**Comment by vpisarev** [comment]
@Developer-Ecosystem-Engineering, thank you! the patch is almost ready to be merged. Please, fix the compile warnings on Windows (see pullrequest.opencv.org) and squash commits into one.

**Comment by Developer-Ecosystem-Engineering** [comment]
> @Developer-Ecosystem-Engineering, thank you! the patch is almost ready to be merged. Please, fix the compile warnings on Windows (see pullrequest.opencv.org) and squash commits into one.

- Squashed into a single commit
- Fix windows build errors
- Use OpenCV universal intrinsics
- Remove CV_NEON/CV_SIMD guards

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

File: modules/features2d/src/sift.simd.hpp [hunk 3]
```diff
@@ -450,31 +450,182 @@ class findScaleSpaceExtremaT
...
+                mask &= v_signmask(condp | condm);
+
+                if (mask == 0)
+                {
+                    continue;
+                }
+
+                for (int k = 0; k<vecsize;k++) if (mask & (1<<k))
+                {
+                    CV_TRACE_REGION("pixel_candidate");
```
[the remark this answers is not in the document]
Comment (Developer-Ecosystem-Engineering) [reply modules/features2d/src/sift.simd.hpp]: Resolved in latest.

File: modules/features2d/src/sift.simd.hpp [hunk 4]
```diff
@@ -450,31 +450,182 @@ class findScaleSpaceExtremaT
...
+                condm = (val <= vmin) & (val <= min_middle);
+
+                mask &= v_signmask(condp | condm);
+
+                if (mask == 0)
+                {
+                    continue;
+                }
+
+                for (int k = 0; k<vecsize;k++) if (mask & (1<<k))
```
[the remark this answers is not in the document]
Comment (Developer-Ecosystem-Engineering) [reply modules/features2d/src/sift.simd.hpp]: Resolved in latest.

File: modules/features2d/src/sift.simd.hpp [hunk 5]
```diff
@@ -450,31 +450,182 @@ class findScaleSpaceExtremaT
...
+                _10 = vx_load(&nextptr[c     -1]);                                  _12 = vx_load(&nextptr[c     +1]);
+                _20 = vx_load(&nextptr[c+step-1]); _21 = vx_load(&nextptr[c+step]); _22 = vx_load(&nextptr[c+step+1]);
+
+                vmax = v_max(v_max(v_max(_00,_01),v_max(_02,_10)),v_max(v_max(_12,_20),v_max(_21,_22)));
+                vmin = v_min(v_min(v_min(_00,_01),v_min(_02,_10)),v_min(v_min(_12,_20),v_min(_21,_22)));
+
+                condp = (val >= vmax) & (val >= max_middle);
+                condm = (val <= vmin) & (val <= min_middle);
+
+                mask &= v_signmask(condp | condm);
```
[the remark this answers is not in the document]
Comment (Developer-Ecosystem-Engineering) [reply modules/features2d/src/sift.simd.hpp]: Resolved in latest.

File: modules/features2d/src/sift.simd.hpp [hunk 6]
```diff
@@ -450,31 +450,182 @@ class findScaleSpaceExtremaT
...
+                v_float32 val = vx_load(&currptr[c]);
+                v_float32 _00,_01,_02;
+                v_float32 _10,    _12;
+                v_float32 _20,_21,_22;
+
+                v_float32 vmin,vmax;
+
+
+                v_float32 cond = v_abs(val) > vx_setall_f32((float)threshold);
+                if (v_signmask(cond) == 0)
```
[the remark this answers is not in the document]
Comment (Developer-Ecosystem-Engineering) [reply modules/features2d/src/sift.simd.hpp]: Resolved in latest.

**Comment by Developer-Ecosystem-Engineering** [comment]
Modifications requested by @alalek have been integrated and re-squashed.

**Review by alalek — APPROVED** [review]
Looks good to me :+1:

/cc @vpisarev

**Pull request merged**